File: src/mosh/ops.ts
// Pure datamosh operations over a flat list of AVI frame chunks. No ffmpeg here,
// so these are deterministic and unit-testable.
//
// A useful free signal: a P-frame chunk's byte length (`data.length`) tracks how
// much it changed from the previous frame — i.e. roughly its motion/residual
// magnitude. Several effects below exploit that (strongest-frame bloom, sort).
import type { AviChunk } from './avi'
import { frameType } from './avi'

const DEFAULT_SEED = 0x6d6f7368 // "mosh"

function clamp01(x: number): number {
  return Math.max(0, Math.min(1, x))
}

function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * clamp01(t)
}

/** Split off the leading seed keyframe; return it plus the trailing frames. */
function splitSeed(chunks: AviChunk[]): { seed: AviChunk[]; rest: AviChunk[] } {
  if (chunks.length === 0) return { seed: [], rest: [] }
  return { seed: [chunks[0]], rest: chunks.slice(1) }
}

/** Drop interior keyframes (keep the first), so motion bleeds across the frame. */
export function bloom(chunks: AviChunk[], intensity = 1, seed = DEFAULT_SEED): AviChunk[] {
  const p = clamp01(intensity)
  const rng = mulberry32(seed)
  let seenFirst = false
  return chunks.filter((c) => {
    if (frameType(c.data) !== 'I') return true
    if (!seenFirst) {
      seenFirst = true
      return true
    }
    return rng() >= p
  })
}

/**
 * Stutter / P-frame repeat: periodically duplicate predicted frames so their
 * motion vectors re-apply, freezing/echoing motion into a glitchy judder.
 */
export function stutter(chunks: AviChunk[], intensity = 0.5): AviChunk[] {
  const t = clamp01(intensity)
  if (t <= 0) return chunks.slice()
  const period = Math.max(2, Math.round(16 * (1 - t)))
  const extra = Math.max(1, Math.round(8 * t))

  const out: AviChunk[] = []
  let pIndex = 0
  for (const c of chunks) {
    out.push(c)
    if (frameType(c.data) === 'P') {
      if (pIndex % period === 0) {
        for (let i = 0; i < extra; i++) out.push(c)
      }
      pIndex += 1
    }
  }
  return out
}

/**
 * Bloom burst: drop interior keyframes, then find the single strongest-motion
 * P-frame (largest chunk) and repeat it many times in place — the explosive,
 * psychedelic bloom. `intensity` scales the repeat count (~4–48).
 */
export function bloomBurst(chunks: AviChunk[], intensity = 0.7, seed = DEFAULT_SEED): AviChunk[] {
  const bloomed = bloom(chunks, 1, seed)
  // Index of the strongest P-frame.
  let best = -1
  let bestLen = -1
  for (let i = 0; i < bloomed.length; i++) {
    if (frameType(bloomed[i].data) === 'P' && bloomed[i].data.length > bestLen) {
      bestLen = bloomed[i].data.length
      best = i
    }
  }
  if (best < 0) return bloomed
  const copies = Math.round(lerp(4, 48, intensity))
  const burst = Array.from({ length: copies }, () => bloomed[best])
  return [...bloomed.slice(0, best + 1), ...burst, ...bloomed.slice(best + 1)]
}

/**
 * Shuffle: drop interior keyframes, then shuffle the predicted frames in blocks
 * so motion is re-applied out of order — chaotic scramble. Higher intensity uses
 * smaller blocks (finer chaos).
 */
export function shuffle(chunks: AviChunk[], intensity = 0.5, seed = DEFAULT_SEED): AviChunk[] {
  const bloomed = bloom(chunks, 1, seed)
  const { seed: head, rest } = splitSeed(bloomed)
  const block = Math.max(1, Math.round(lerp(8, 1, intensity)))
  const blocks: AviChunk[][] = []
  for (let i = 0; i < rest.length; i += block) blocks.push(rest.slice(i, i + block))
  const rng = mulberry32(seed ^ 0x9e3779b9)
  for (let i = blocks.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1))
    ;[blocks[i], blocks[j]] = [blocks[j], blocks[i]]
  }
  return [...head, ...blocks.flat()]
}

/**
 * Sort: drop interior keyframes, then order the predicted frames by motion
 * magnitude (strongest first) so the smear builds and resolves unnaturally.
 */
export function sortByMotion(chunks: AviChunk[]): AviChunk[] {
  const bloomed = bloom(chunks, 1)
  const { seed: head, rest } = splitSeed(bloomed)
  const sorted = rest.slice().sort((a, b) => b.data.length - a.data.length)
  return [...head, ...sorted]
}

/**
 * Reverse: drop interior keyframes, then reverse the predicted frames so motion
 * plays backward over forward content.
 */
export function reverse(chunks: AviChunk[]): AviChunk[] {
  const bloomed = bloom(chunks, 1)
  const { seed: head, rest } = splitSeed(bloomed)
  return [...head, ...rest.reverse()]
}

/**
 * Transition bloom: concatenate clip A and clip B, dropping B's leading
 * keyframe(s) at the cut so A's final motion bleeds into B.
 */
export function transition(a: AviChunk[], b: AviChunk[]): AviChunk[] {
  let i = 0
  while (i < b.length && frameType(b[i].data) === 'I') i += 1
  return [...a, ...b.slice(i)]
}

/** A fractional [start, end] window over a clip (0..1). */
export interface Range {
  start: number
  end: number
}

export function isFullRange(range?: Range): boolean {
  return !range || (range.start <= 0 && range.end >= 1)
}

/**
 * Apply a frame transform to only a sub-window of the clip. Frames before/after
 * the window pass through untouched, so the moshed region is bracketed by clean
 * footage — the "mosh from here to there" workflow. The caller is responsible
 * for transcoding with keyframes inside `after` so the picture re-establishes.
 */
export function applyWindowed(
  transform: (chunks: AviChunk[]) => AviChunk[],
  chunks: AviChunk[],
  range?: Range,
): AviChunk[] {
  if (isFullRange(range)) return transform(chunks)
  const n = chunks.length
  const clamp = (v: number) => Math.max(0, Math.min(n, Math.round(v * n)))
  const s = clamp(range!.start)
  const e = Math.max(s, clamp(range!.end))
  const before = chunks.slice(0, s)
  const window = chunks.slice(s, e)
  const after = chunks.slice(e)
  return [...before, ...transform(window), ...after]
}

/** Tiny deterministic PRNG (mulberry32) returning floats in [0, 1). */
function mulberry32(seed: number): () => number {
  let s = seed >>> 0
  return () => {
    s = (s + 0x6d2b79f5) >>> 0
    let t = s
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

File: crates/dmosh-core/src/ops.rs
//! Pure datamosh operations over a flat list of AVI frame chunks — a port of
//! `src/mosh/ops.ts`. Chunk cloning is cheap (the data is `Rc`-shared).

use crate::avi::{frame_type, Chunk};

#[derive(Clone, Copy)]
pub struct DropTargets {
    pub i: bool,
    pub p: bool,
}
pub const DEFAULT_TARGETS: DropTargets = DropTargets { i: true, p: false };

fn clamp01(x: f32) -> f32 {
    x.max(0.0).min(1.0)
}
fn lerp(a: f32, b: f32, t: f32) -> f32 {
    a + (b - a) * clamp01(t)
}

fn mulberry32(seed: u32) -> impl FnMut() -> f64 {
    let mut s = seed;
    move || {
        s = s.wrapping_add(0x6d2b79f5);
        let mut t = s;
        t = (t ^ (t >> 15)).wrapping_mul(t | 1);
        t ^= t.wrapping_add((t ^ (t >> 7)).wrapping_mul(t | 61));
        ((t ^ (t >> 14)) as f64) / 4294967296.0
    }
}

fn split_seed(chunks: &[Chunk]) -> (Vec<Chunk>, Vec<Chunk>) {
    if chunks.is_empty() {
        return (vec![], vec![]);
    }
    (vec![chunks[0].clone()], chunks[1..].to_vec())
}

/// Keep the first keyframe as a seed, drop every later keyframe.
pub fn strip_interior_keyframes(chunks: &[Chunk]) -> Vec<Chunk> {
    let mut seen = false;
    chunks
        .iter()
        .filter(|c| {
            if frame_type(&c.data) != b'I' {
                return true;
            }
            if !seen {
                seen = true;
                return true;
            }
            false
        })
        .cloned()
        .collect()
}

/// Bloom: duplicate every predicted frame so its motion re-applies — the
/// continuous melt. `intensity` sets extra copies per P-frame (1..4).
pub fn bloom(chunks: &[Chunk], intensity: f32) -> Vec<Chunk> {
    let copies = 1 + (clamp01(intensity) * 3.0).round() as usize;
    let mut out = Vec::with_capacity(chunks.len() * (copies + 1));
    for c in chunks {
        out.push(c.clone());
        if frame_type(&c.data) == b'P' {
            for _ in 0..copies {
                out.push(c.clone());
            }
        }
    }
    out
}

/// Stutter: periodically repeat predicted frames. `intensity` is the rate;
/// `repeat` (optional) the explicit copies per hit.
pub fn stutter(chunks: &[Chunk], intensity: f32, repeat: Option<u32>) -> Vec<Chunk> {
    let t = clamp01(intensity);
    if t <= 0.0 {
        return chunks.to_vec();
    }
    let period = ((16.0 * (1.0 - t)).round() as usize).max(1);
    let extra = repeat
        .map(|r| (r as usize).max(1))
        .unwrap_or_else(|| ((8.0 * t).round() as usize).max(1));

    let mut out = Vec::new();
    let mut p_index = 0usize;
    for c in chunks {
        out.push(c.clone());
        if frame_type(&c.data) == b'P' {
            if p_index % period == 0 {
                for _ in 0..extra {
                    out.push(c.clone());
                }
            }
            p_index += 1;
        }
    }
    out
}

/// Bloom burst: repeat the single strongest-motion P-frame many times.
pub fn bloom_burst(chunks: &[Chunk], intensity: f32) -> Vec<Chunk> {
    let mut best: isize = -1;
    let mut best_len = 0usize;
    for (i, c) in chunks.iter().enumerate() {
        if frame_type(&c.data) == b'P' && c.data.len() > best_len {
            best_len = c.data.len();
            best = i as isize;
        }
    }
    if best < 0 {
        return chunks.to_vec();
    }
    let best = best as usize;
    let copies = lerp(8.0, 50.0, intensity).round() as usize;
    let mut out = Vec::with_capacity(chunks.len() + copies);
    out.extend_from_slice(&chunks[..=best]);
    for _ in 0..copies {
        out.push(chunks[best].clone());
    }
    out.extend_from_slice(&chunks[best + 1..]);
    out
}

/// Shuffle: drop interior keyframes, then shuffle predicted frames in blocks.
pub fn shuffle(chunks: &[Chunk], intensity: f32, seed: u32) -> Vec<Chunk> {
    let stripped = strip_interior_keyframes(chunks);
    let (head, rest) = split_seed(&stripped);
    let block = (lerp(8.0, 1.0, intensity).round() as usize).max(1);
    let mut blocks: Vec<Vec<Chunk>> = rest.chunks(block).map(|b| b.to_vec()).collect();
    let mut rng = mulberry32(seed ^ 0x9e37_79b9);
    let mut i = blocks.len();
    while i > 1 {
        i -= 1;
        let j = (rng() * (i as f64 + 1.0)).floor() as usize;
        blocks.swap(i, j);
    }
    let mut out = head;
    for b in blocks {
        out.extend(b);
    }
    out
}

/// Sort: drop interior keyframes, then order predicted frames by motion energy.
pub fn sort_by_motion(chunks: &[Chunk]) -> Vec<Chunk> {
    let stripped = strip_interior_keyframes(chunks);
    let (head, mut rest) = split_seed(&stripped);
    rest.sort_by(|a, b| b.data.len().cmp(&a.data.len()));
    let mut out = head;
    out.extend(rest);
    out
}

/// Reverse: drop interior keyframes, then reverse predicted frames.
pub fn reverse(chunks: &[Chunk]) -> Vec<Chunk> {
    let stripped = strip_interior_keyframes(chunks);
    let (head, mut rest) = split_seed(&stripped);
    rest.reverse();
    let mut out = head;
    out.extend(rest);
    out
}

/// Transition: concat A + B, stripping targeted frame types from a leading
/// window of B (length `bleed`) so A melts into B.
pub fn transition(a: &[Chunk], b: &[Chunk], targets: DropTargets, bleed: f32) -> Vec<Chunk> {
    let window_len = ((clamp01(bleed) * b.len() as f32).round() as usize).max(1);
    let mut out = a.to_vec();
    for (i, c) in b.iter().enumerate() {
        let t = frame_type(&c.data);
        let in_window = i < window_len;
        let drop = in_window && ((t == b'I' && targets.i) || (t == b'P' && targets.p));
        if !drop {
            out.push(c.clone());
        }
    }
    out
}

/// Apply a transform to only a fractional sub-window of the clip.
pub fn apply_windowed<F>(transform: F, chunks: &[Chunk], range: Option<(f32, f32)>) -> Vec<Chunk>
where
    F: Fn(&[Chunk]) -> Vec<Chunk>,
{
    let (start, end) = match range {
        Some((s, e)) if !(s <= 0.0 && e >= 1.0) => (s, e),
        _ => return transform(chunks),
    };
    let n = chunks.len();
    let clamp = |v: f32| ((v * n as f32).round() as usize).min(n);
    let s = clamp(start);
    let e = clamp(end).max(s);
    let mut out = chunks[..s].to_vec();
    out.extend(transform(&chunks[s..e]));
    out.extend_from_slice(&chunks[e..]);
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::avi::frame_type;
    use std::rc::Rc;

    fn f(t: u8, payload: usize) -> Chunk {
        let ct = if t == b'I' { 0x00 } else { 0x40 };
        let mut data = vec![0u8, 0, 1, 0xb6, ct];
        data.extend((0..payload).map(|i| (i + 1) as u8));
        Chunk { id: *b"00dc", data: Rc::from(data.as_slice()) }
    }
    fn stream(p: &str) -> Vec<Chunk> {
        p.bytes().map(|b| f(b, 4)).collect()
    }
    fn count_i(c: &[Chunk]) -> usize {
        c.iter().filter(|x| frame_type(&x.data) == b'I').count()
    }

    #[test]
    fn bloom_duplicates_p_frames() {
        let out = bloom(&stream("IPPP"), 1.0); // 4 extra each → P becomes 5
        assert_eq!(out.len(), 1 + 3 * 5);
        assert_eq!(count_i(&out), 1);
    }

    #[test]
    fn stutter_repeat_count() {
        let out = stutter(&stream("IPPP"), 1.0, Some(3)); // every P, +3
        assert_eq!(out.len(), 1 + 3 * 4);
        assert_eq!(count_i(&out), 1);
    }

    #[test]
    fn transition_bleed() {
        let a = stream("IPP");
        let b = stream("IPPIPP"); // keyframes at 0 and 3
        let cut = transition(&a, &b, DEFAULT_TARGETS, 0.0);
        let types: String = cut.iter().map(|c| frame_type(&c.data) as char).collect();
        assert_eq!(types, "IPPPPIPP");
        let full = transition(&a, &b, DEFAULT_TARGETS, 1.0);
        let types: String = full.iter().map(|c| frame_type(&c.data) as char).collect();
        assert_eq!(types, "IPPPPPP");
    }

    #[test]
    fn windowed_only_touches_window() {
        let drop_all = |cs: &[Chunk]| -> Vec<Chunk> {
            cs.iter().filter(|c| frame_type(&c.data) != b'I').cloned().collect()
        };
        let chunks = stream("IPPIPPIPP");
        let out = apply_windowed(drop_all, &chunks, Some((3.0 / 9.0, 6.0 / 9.0)));
        let types: String = out.iter().map(|c| frame_type(&c.data) as char).collect();
        assert_eq!(types, "IPPPPIPP");
    }
}
